packages/sitecore-jss-nextjs/src/middleware/middleware.ts

```diff
@@ -0,0 +1,103 @@
+   * @param {NextResponse} [res] response object from middleware handler
+   */
+  disabled?: (req?: NextRequest, res?: NextResponse) => boolean;
+  /**
+   * Function used to determine if route should be excluded.
+   * By default, files (pathname.includes('.')), Next.js API routes (pathname.startsWith('/api/')), and Sitecore API routes (pathname.startsWith('/sitecore/')) are ignored.
+   * This is an important performance consideration since Next.js Edge middleware runs on every request.
+   * @param {string} pathname The pathname
+   * @returns {boolean} Whether to exclude the route
+   */
+  excludeRoute?: (pathname: string) => boolean;
+  /**
+   * Fallback hostname in case `host` header is not present
+   * @default localhost
+   */
+  defaultHostname?: string;
+  /**
+   * Site resolution implementation by name/hostname
+   */
+  siteResolver: SiteResolver;
+};
+
+export abstract class MiddlewareBase {
+  protected SITE_SYMBOL = 'sc_site';
+  protected defaultHostname: string;
+
+  constructor(protected config: MiddlewareBaseConfig) {
+    this.defaultHostname = config.defaultHostname || 'localhost';
+  }
+
+  /**
+   * Determines if mode is preview
+   * @param {NextRequest} req request
+   * @returns {boolean} is preview
+   */
+  protected isPreview(req: NextRequest) {
+    return !!(
+      req.cookies.get('__prerender_bypass')?.value || req.cookies.get('__next_preview_data')?.value
+    );
+  }
+
+  protected excludeRoute(pathname: string) {
+    return (
+      pathname.includes('.') || // Ignore files
+      pathname.startsWith('/api/') || // Ignore Next.js API calls
+      pathname.startsWith('/sitecore/') || // Ignore Sitecore API calls
+      pathname.startsWith('/_next') || // Ignore next service calls
+      (this.config?.excludeRoute && this.config?.excludeRoute(pathname))
+    );
+  }
+
+  /**
+   * Safely extract all headers for debug logging
+   * Necessary to avoid middleware issue https://github.com/vercel/next.js/issues/39765
+   * @param {Headers} incomingHeaders Incoming headers
+   * @returns Object with headers as key/value pairs
+   */
+  protected extractDebugHeaders(incomingHeaders: Headers) {
+    const headers = {} as { [key: string]: string };
+    incomingHeaders.forEach((value, key) => (headers[key] = value));
+    return headers;
+  }
+
+  /**
+   * Provides used language
+   * @param {NextRequest} req request
+   * @returns {string} language
+   */
+  protected getLanguage(req: NextRequest) {
+    return req.nextUrl.locale || req.nextUrl.defaultLocale || 'en';
+  }
+
+  /**
+   * Extract 'host' header
+   * @param {NextRequest} req request
+   */
+  protected getHostHeader(req: NextRequest) {
+    return req.headers.get('host')?.split(':')[0];
+  }
+
+  /**
+   * Get site information
+   * @param {NextRequest} req request
+   * @param {NextResponse} [res] response
+   * @returns {SiteInfo} site information
+   */
+  protected getSite(req: NextRequest, res?: NextResponse): SiteInfo {
```

Comment:
Do we want to put some safeguard in this function to ensure `siteResolver` is not used in preview mode? Either check `isPreview` here - or at least put something in the comment so devs remember not to use this function if preview mode.

Reply:
Didn't decide to check _isPreview_ here, because in this case, it requires returning _undefined_, I don't want to have an optional result. Added additional sentence to description